### PR TITLE
Fixed some undefined behavior and added CodeQL workflow

### PR DESCRIPTION
**Pull request by weichslgartner**
### Issue
```
mkdir build && cd build
CC=clang cmake -DCMAKE_C_FLAGS="-fsanitize=address,undefined" -DCMAKE_EXE_LINKER_FLAGS="-fsanitize=address,undefined" .. && cmake --build .
```
```
$echo "(0.0) can1 ffffffff#00000000" | ./log2long 
can-utils/lib.c:187:23: runtime error: left shift of 15 by 28 places cannot be represented in type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /can-utils/lib.c:187:23 in 
(0.0)  can1  3FFFFFFF   [4]  remote request
```

### Proposed fix

cast to unsigned value
`cf->can_id |= ((canid_t)tmp << (7-i)*4);`

### Issue
```
mkdir build && cd build
CC=clang cmake -DCMAKE_C_FLAGS="-fsanitize=address,undefined" -DCMAKE_EXE_LINKER_FLAGS="-fsanitize=address,undefined" .. && cmake --build .
```
```
$echo "echo "0.0000 0  Rx   d 8 8D 00 10 01 00 82 01 00                                                                                    
      0.200000000000000000 0- 0000  Rx   d 8 8D 00 10 01 00 82 01 00" | ./asc2log
can-utils/asc2log.c:100:20: runtime error: signed integer overflow: 200000000000000000 * 100 cannot be represented in type 'long'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior asc2log.c:100:20 in 
```

### Proposed fix
Add overflow check before performing the multiplication
```
if (dplace == 4 && read_tv.tv_usec >= INT_MAX/100)
    return;
```

### Issue:
CodeQL found four  occurence of "potentially overflowing call to snprintf" (_The return value of a call to snprintf is the number of characters that would have been written to the buffer assuming there was sufficient space. In the event that the operation reaches the end of the buffer and more than one character is discarded, the return value will be greater than the buffer size._)

### Proposed fix
Check return value of snprintf:
```
tmp_n = snprintf(buf + n, len - n, ",");
if (tmp_n < 0 || tmp_n >= len - n){
	return n;
}
n += tmp_n;
```

### CODE REVIEW

File: asc2log.c [hunk 1]
```diff
@@ -264,6 +272,10 @@ void eval_canfd(char* buf, struct timeval *date_tvp, char timestamps, int dplace
 	if (strlen(dir) != 2) /* "Rx" or "Tx" */
 		return;
 
+	/* check for signed integer overflow */
+	if (dplace == 4 && read_tv.tv_usec >= INT_MAX/100)
+		return;
+
```
Comment (hartkopp) [inline asc2log.c:278]: No check for dplace == 5 as in eval_can() ?

File: asc2log.c [hunk 2]
```diff
@@ -412,7 +424,7 @@ int main(int argc, char **argv)
 	static char base; /* 'd'ec or 'h'ex */
 	static char timestamps; /* 'a'bsolute or 'r'elative */
 	int opt;
-
+	//while (__AFL_LOOP(10000)){
```
Comment (hartkopp) [inline asc2log.c]: Debug leftover?

File: lib.c [hunk 3]
```diff
@@ -184,6 +184,8 @@ int parse_canframe(char *cs, struct canfd_frame *cf) {
 		for (i=0; i<8; i++){
 			if ((tmp = asc2nibble(cs[i])) > 0x0F)
 				return 0;
+
+			/* casting tmp to canid_t to perform left shift of 28 */
```
Comment (hartkopp) [inline lib.c:188]: Would it make sense to make tmp a canid_t ?

File: asc2log.c (hunk 1 above)
Comment (weichslgartner) [reply asc2log.c:278]: Yes seems that I forgot this.

File: asc2log.c (hunk 2 above)
Comment (weichslgartner) [reply asc2log.c]: Yes, can be deleted.

File: asc2log.c (hunk 1 above)
Comment (weichslgartner) [reply asc2log.c:278]: Should be there now

File: lib.c (hunk 3 above)
Comment (weichslgartner) [reply lib.c]: Done

**Comment by weichslgartner** [comment]
Updated the PR.

**Comment by weichslgartner** [comment]
@hartkopp what is the status here? Can I do anything?

**Comment by hartkopp** [comment]
Oh, sorry about that!
I'm fine with the code changes but I would like to have some feedback from @marckleinebudde who is more experienced in the GitHub CI and probably CodeQL workflows.

**Comment by marckleinebudde** [comment]
You do you add the various workflow file and delete them again?
Can you squash the patches so that each problem is fixed in one patch? E.g. the `can_id` overflow ine one patch the `snprintf` in another, etc...

**Comment by weichslgartner** [comment]
@marckleinebudde  the other workflow files produced too much false positives or needed additional work. That is why I removed them. I adapted now the changes into commits to the specific fixes/additions.

File: asc2log.c [hunk 4]
```diff
@@ -182,6 +183,13 @@ void eval_can(char* buf, struct timeval *date_tvp, char timestamps, char base, i
 		if (strlen(dir) != 2) /* "Rx" or "Tx" */
 			return;
 
+		/* check for signed integer overflow */
+		if (dplace == 4 && read_tv.tv_usec >= INT_MAX/100)
```
Comment (marckleinebudde) [inline asc2log.c:187]: please add spaces areound the `/`

Comment (marckleinebudde) [reply asc2log.c:187]: Does it make sense to move the overflow check into `calc_tv()`?

File: asc2log.c [hunk 5]
```diff
@@ -412,7 +428,6 @@ int main(int argc, char **argv)
 	static char base; /* 'd'ec or 'h'ex */
 	static char timestamps; /* 'a'bsolute or 'r'elative */
 	int opt;
-
```
Comment (marckleinebudde) [inline asc2log.c:415]: unrelated change

**Review by marckleinebudde — CHANGES_REQUESTED** [review]

**Review by marckleinebudde — CHANGES_REQUESTED** [review]

**Comment by marckleinebudde** [comment]
> @marckleinebudde the other workflow files produced too much false positives or needed additional work. That is why I removed them.

I see, makes sense.

> I adapted now the changes into commits to the specific fixes/additions.

Now reviewing is easier. Looks almost ready. Can you rebase to the current `master` branch?.

File: asc2log.c (hunk 4 above)
Comment (weichslgartner) [reply asc2log.c]: added spaces. If check is moved to calc_tv then the struct musst be set to a special value and checked later if time struct is valid, or? Seems to be easier to return before calulating invalid tv.

File: asc2log.c (hunk 5 above)
Comment (weichslgartner) [reply asc2log.c]: deleted empty line

**Comment by weichslgartner** [comment]
done the changes and rebased to upstream master

File: lib.c [hunk 6]
```diff
@@ -159,7 +159,6 @@ int parse_canframe(char *cs, struct canfd_frame *cf) {
 	int i, idx, dlen, len;
 	int maxdlen = CAN_MAX_DLEN;
 	int ret = CAN_MTU;
-	/* unsigned type to prevent signed integer shifting and undefined behavior */
```
Comment (marckleinebudde) [inline lib.c]: This should go into the previous patch.

**Review by marckleinebudde — CHANGES_REQUESTED** [review]

File: lib.c (hunk 6 above)
Comment (marckleinebudde) [reply lib.c]: please squash the new patch
